zeus/device/cpu/rapl.py

```diff
@@ -261,6 +265,65 @@ def supportsGetDramEnergyConsumption(self) -> bool:
         return self.dram is not None
 
 
+class ZeusdRAPLCPU(RAPLCPU):
+    """A RAPLCPU that interfaces with RAPL via zeusd.
+
+    The parent RAPLCPU class requires root privileges to interface with RAPL.
+    ZeusdRAPLCPU (this class) overrides RAPLCPU's methods so that they instead send requests
+    to the Zeus daemon, which will interface with RAPL on behalf of ZeusdRAPLCPU. As a result,
+    ZeusdRAPLCPU does not need root privileges to monitor CPU and DRAM energy consumption.
+
+    See [here](https://ml.energy/zeus/getting_started/#system-privileges)
+    for details on system privileges required.
+    """
+
+    def __init__(
+        self,
+        cpu_index: int,
+        zeusd_sock_path: str = "/var/run/zeusd.sock",
+    ) -> None:
+        """Initialize the Intel CPU with a specified index."""
+        self.cpu_index = cpu_index
+
+        self._client = httpx.Client(transport=httpx.HTTPTransport(uds=zeusd_sock_path))
+        self._url_prefix = f"http://zeusd/cpu/{cpu_index}"
+
+    def getTotalEnergyConsumption(self) -> CpuDramMeasurement:
+        """Returns the total energy consumption of the specified powerzone. Units: mJ."""
+        resp = self._client.post(
+            self._url_prefix + "/get_index_energy",
+            json={
+                "cpu": True,
+                "dram": True,
+            },
+        )
+        if resp.status_code != 200:
+            raise ZeusdError(f"Failed to get total energy consumption: {resp.text}")
+
+        data = resp.json()
+        cpu_uj = data.get("cpu_energy_uj")
+        dram_uj = data.get("dram_energy_uj")
+        cpu_mj = None if cpu_uj is None else cpu_uj / 1000
+        dram_mj = None if dram_uj is None else dram_uj / 1000
```

Comment:
@wbjin Quick question -- Here we've constructed the request with `cpu` and `dram` both set to `True`. In that case, can the response `cpu_energy_uj` ever be `None`? Because if there was some error from the daemon-side, the error would be propagated back to the Python client, so the status code won't be 200. On the other hand, `dram` may actually be `None` because the underlying RAPL device may not support DRAM measurement. Just wanted to confirm this with you.

Assuming that the above is true, `cpu_energy_uj` somehow ending up as a `None` means something went wrong. Zeus should error out.
```suggestion
        cpu_mj = data["cpu_energy_uj"] / 1000
        
        dram_uj = data.get("dram_energy_uj")
        dram_mj = dram_uj / 1000 if dram_uj is not None else None
```

Reply:
Thank you for pointing that out -- this was my mistake. I don't think `cpu_energy_uj` can ever be None here if zeusd is working correctly. But @wbjin please confirm if this sounds right!

Reply:
We might go even further here. In the constructor, we can call `supportsGetDramEnergyConsumption` (via the daemon) once and cache the result inside the class. Subsequent calls to `supportsGetDramEnergyConsumption` will return the cached value. Also, in `getTotalEnergyConsumption`, we would be able to know whether `None` is a valid response value for the `dram` field.

Reply:
Yup, if `cpu` and `dram` are set to be both `True` in the request, `cpu_energy_uj` shouldn't be `None` while `dram_energy_uj` can be `None` if dram isn't supported.